ArduPilot/ardupilot
# Use AP_Arming ubiquitously in vehicles for vehicle arming

Mostly just namespace changes from (e.g.) `Copter::init_arm_motors` to `copter.arming.arm()`, but brings the vehicles a lot closer together in terms of their arming.

Some of these commits could possibly be squashed.  Several of them are there to preserve behaviour - e.g. Plane emitting "Throttle armed" as a statustext but Copter not.

Changes in behaviour:

- MAVLink: Rover gets force-disarm and force-arm via mavlink
- MAVLink: Sub gets force-disarm and force-arm via mavlink
- MAVLink: Plane gets force-disarm
- Canonicalise on "if you try to arm via mavlink when armed then return success", similarly for disarm
- arming method for Rover's aux-channel arming corrected from rudder to auxswitch
 - BOARD_SAFETY_OPTION_SAFETY_ON_DISARM honoured on all vehicles as it is in the base class method

Allows future cleanups:
 - AP_Arming singleton allows access to vehicle arm state - at least this concept of armed
 - remove set_soft_armed in favour of AP_HAL checking AP_Arming::is_armed and hal.util->safety_switch_state
   - this will require a behavioural change from some vehicles so hal.util's concept of armed is common between our vehicles
   - elimination of update_soft_armed methods
 - moving of proximity prearm check up (so Rover can use it)
 - Copter's AUX_FUNC::ARMDISARM should move up and the AP_Arming::Method stored
 - rudder arm/disarm can possibly move up
 - vehicle arm/disarm logging can become common (probably log both an even and something into the ARM message as the ARM message contains more information)
 - reduce the number of "armed" states we have in the code
 - make reset-mission-on-disarm a common option between vehicles
 - moving of arming-related methods into AP_Arming files
 - have the parachute library itself disarm the vehicle rather than doing that in the vehicles
 - move manipulation of logger armed status into base class (AP::logger().set_vehicle_armed())
 - move AP_Notify manipulation up to base class (so Rover and Plane get arming tones)
 - move AP_Scheduler manipulation up to base class
 - could consider storing armed-bearing in AP_Arming
 - canonicalise home position / baro reset / EKF height datum reset handling
 - make resetting EKF height datum an arming check rather than something we do once armed (as it is bool and is kinda-sorta MUST_CHECK)
 - move sprayer-pump-testing-disabling up to base arm method
 - canonicalise return value of AP_Arming::arm() when we are armed to return false
 - canonicalise return value of AP_Arming::disarm() when we are disarmed to return false
  - very few places actually check the return value
 - consider making in_arming_delay something in the base class and/or renaming it to in_post_arming_delay
 - try to work out of motor estop functionality can be made common in AP_Arming
 - make safety switch behaviour common between vehicles for whether you can arm or not
 - create AP_Arming::update() and continuously do prearms on Plane, Sub and Rover as they are in Copter
 - move inertial-nav vs baro check up to base class(?)
 - remove defaulting in "bool arm(AP_Arming::Method method, bool do_arming_checks=true) override;"
 - consider not allowing arm() to be overridden but have hooks for post-arm actions
